velox/functions/sparksql/Mask.cpp

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <utility>
+
+#include "velox/expression/DecodedArgs.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/expression/VectorWriters.h"
+#include "velox/functions/lib/Re2Functions.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+class MaskFunction final : public exec::VectorFunction {
+  static constexpr std::string_view maskedUpperCase{"X"};
+  static constexpr std::string_view maskedLowerCase{"x"};
+  static constexpr std::string_view maskedDigit{"n"};
+
+ public:
+  MaskFunction() {}
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& /* outputType */,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    // Get the decoded vectors out of arguments.
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    DecodedVector* strings = decodedArgs.at(0);
+    DecodedVector* upperChars = args.size() >= 2 ? decodedArgs.at(1) : nullptr;
+    DecodedVector* lowerChars = args.size() >= 3 ? decodedArgs.at(2) : nullptr;
+    DecodedVector* digitChars = args.size() >= 4 ? decodedArgs.at(3) : nullptr;
+    DecodedVector* otherChars = args.size() >= 5 ? decodedArgs.at(4) : nullptr;
+    BaseVector::ensureWritable(rows, VARCHAR(), context.pool(), result);
+    auto* results = result->as<FlatVector<StringView>>();
+    // Optimization for the (flat, const, const, const, const) case.
+    if (strings->isIdentityMapping() and
+        (upperChars == nullptr || upperChars->isConstantMapping()) and
+        (lowerChars == nullptr || lowerChars->isConstantMapping()) and
+        (digitChars == nullptr || digitChars->isConstantMapping()) and
+        (otherChars == nullptr || otherChars->isConstantMapping())) {
+      // TODO: enable the inpalce if possible
```

Comment:
Could you clarify what does this TODO refers to, and will it be addressed in this PR?
nit: inpalce -> inplace

Reply:
hey rui, it means if the input string can be reused into result output. For example, if input string only have one reference, then instead of constructing new result string instance just replace the input string itself and put it into result.  I just checked how `replace` does by referencing `replace` function support https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/StringFunctions.cpp#L104

I think it more for a memory related optimization, not mandarary, will not address in this PR but a TODO item may can address in seperate PR.